crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2717,6 +2717,21 @@ impl<'db> TypeInferenceBuilder<'db> {
         // - `[ast::CompOp::Is]`: return `false` if unequal, `bool` if equal
         // - `[ast::CompOp::IsNot]`: return `true` if unequal, `bool` if equal
         match (left, right) {
+            (Type::Union(union), other) => {
+                let mut builder = UnionBuilder::new(self.db);
+                for element in union.elements(self.db) {
+                    builder = builder.add(self.infer_binary_type_comparison(*element, op, other)?);
+                }
+                Some(builder.build())
+            }
+            (other, Type::Union(union)) => {
+                let mut builder = UnionBuilder::new(self.db);
+                for element in union.elements(self.db) {
+                    builder = builder.add(self.infer_binary_type_comparison(other, op, *element)?);
+                }
+                Some(builder.build())
```

Comment:
Would be nice to avoid the duplication here, but I couldn't think of a good way to do this without introducing much more code.

I also looked for ways to replace the `for` loop with an iterator expression, but that would require something like [`Iterator::try_collect`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_collect), which is nightly-only. And would require an additional allocation.

Reply:
I don't mind the repetition (or use of a for loop, I'm a huge fan of for loops) but maybe `Itertools::fold_options` is what you want

https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.fold_options

Reply:
> maybe `Itertools::fold_options` is what you want
> 
> https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.fold_options

Hm, not quite. `fold_options` short-circuits on `None`s in the input iterator. But we want to short-circuit if the output of the `fold`-function returns None. So we would need something like [`Itertools::fold_while`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.fold_while). That *works*, but it's completely ridiculous:
```rs
union
    .elements(self.db)
    .iter()
    .fold_while(Some(UnionBuilder::new(self.db)), |builder, element| {
        if let Some(ty) = self.infer_binary_type_comparison(other, op, *element) {
            FoldWhile::Continue(builder.map(|b| b.add(ty)))
        } else {
            FoldWhile::Done(None)
        }
    })
    .into_inner()
    .map(UnionBuilder::build)
```

I'll go with the `for` loop :smile: